utils/managed_bytes.hh

```diff
@@ -72,27 +72,76 @@ struct blob_storage {
     }
 } __attribute__((packed));
 
+struct [[gnu::packed]] medium_blob_storage {
+    using size_type = uint32_t;
+    using char_type = bytes_view::value_type;
+
+    struct [[gnu::packed]] ref_type {
+        medium_blob_storage* ptr = nullptr;
+        size_type size = 0;
+    };
+
+    ref_type* backref;
+    char_type data[];
+
+    medium_blob_storage(ref_type* backref, size_type size) noexcept
+        : backref(backref)
+    {
+        backref->ptr = this;
+        backref->size = size;
+    }
+
+    medium_blob_storage(medium_blob_storage&& o) noexcept
+        : backref(o.backref)
+    {
+        backref->ptr = this;
+        memcpy(data, o.data, backref->size);
+    }
+
+    size_t storage_size() const noexcept {
+        return sizeof(*this) + backref->size;
+    }
+};
+
 // A managed version of "bytes" (can be used with LSA).
 class managed_bytes {
     friend class bytes_ostream;
     static constexpr size_t max_inline_size = 15;
     struct small_blob {
         bytes_view::value_type data[max_inline_size];
         int8_t size; // -1 -> use blob_storage
+                     // -2 -> use medium_blob_storage
```

Comment:
I thinks `size` should be hoisted out of the `small_blob`, and then we wouldn't need to play games with type punning.
Also a nit: I find it counter-intuitive that we have medium, big and tiny (inline), in other words I would expect big to get the size of -2 and medium the size of -1.
Also we could take this opportunity to use better naming: inline, single-chunk and multi-chunk, instead of opaque big and medium.

Reply:
> Also a nit: I find it counter-intuitive that we have medium, big and tiny (inline), in other words I would expect big to get the size of -2 and medium the size of -1.

The next (cleanup) patch swaps them, and the end result is as you say.

> Also we could take this opportunity to use better naming: inline, single-chunk and multi-chunk, instead of opaque big and medium.

Alright, will do.

> I thinks size should be hoisted out of the small_blob, and then we wouldn't need to play games with type punning.

The next patch puts it at the beginning of all 3 structs, and at that point it's legal because of "common initial sequence" rules.
It could be hoisted out, but that doesn't really change anything at all.

Reply:
> It could be hoisted out, but that doesn't really change anything at all.

It does change things: we no longer need to have these tricky `padding` fields, the usage of which seems very fragile to me. The simpler, the better.

Reply:
We may not even need wrapper structs any more (maybe for documentation purposes only).

Reply:
> It does change things: we no longer need to have these tricky padding fields, the usage of which seems very fragile to me.

The `padding` is just padding. They aren't needed for correctness. I added a comment there explaining that it's only there to make the alignment more natural, in a (very faint) hope that it helps the compiler slightly. (E.g. it puts both `medium` and `big` pointers in the same place, so the instruction which reads it can be common for both branches).

By the way, we can put the padding bytes to work. I have an idea that we could use it to store an allocator signature (e.g. a LSA vs nonLSA bit, or a LSA region ID), for hardening against alloc/free mismatch bugs (https://github.com/scylladb/scylladb/issues/13216).

> We may not even need wrapper structs any more (maybe for documentation purposes only).

I think having a wrapper struct for each variant is nice in any usage of union. I don't see merit in avoiding the wrappers.

Reply:
> The padding fields are just padding. They aren't needed for correctness.

Well, if you add too much of it, the size of the union will be bloated.

Reply:
If you want to use the padding fields, just add regular fields instead.

Reply:
That's why there is exactly enough padding to make `big` and `medium` as big as `small`.

Reply:
> If you want to use the padding fields, just add regular fields instead.

Yes. My point was: you need a wrapper struct for that.

Reply:
I don't mind the wrapper structs. The changes I requested are:
* Hoist size into the `managed_bytes` class itself.
* Remove the padding fields, which have no role anymore.

Reply:
> I don't mind the wrapper structs. The changes I requested are:
> 
>     * Hoist size into the `managed_bytes` class itself.
> 
>     * Remove the padding fields, which have no role anymore.

Alright.

Reply:
It may be difficult to separate size out. We end up with a 15-byte structure, which the compilers don't like to generate. Maybe [[gnu::packed]] helps there.

Reply:
> It may be difficult to separate size out. We end up with a 15-byte structure, which the compilers don't like to generate. Maybe [[gnu::packed]] helps there.

Yes, struct packing reduces its alignment to 1. (It would be weird if it didn't).

Reply:
Any perf implications due to packing?

Reply:
> Any perf implications due to packing?

I'm not sure what you are asking about. Unaligned accesses? These are supposed to have no penalty on modern CPUs.

Reply:
Yes, unaligned. It has no penalty as long as it's all in the same cacheline, no?

Reply:
> Yes, unaligned. It has no penalty as long as it's all in the same cacheline, no?

Even if they are in different cachelines.

https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/

There is some penalty when it crosses page boundaries, though. 

Anyway, I don't think any of this matters. Scylla lives in a different order of magnitude.